test/conformance/ingressv2/util.go

```diff
@@ -720,18 +706,107 @@ func createHTTPRouteReadyDialContext(ctx context.Context, t *testing.T, clients
+	for _, d := range domains {
+		if strings.HasSuffix(string(d), netTest.NetworkingFlags.ClusterSuffix) {
+			return true
+		}
+	}
+	return false
+}
+
+// CreateGateway creates a Gateway.
+func CreateGateway(ctx context.Context, t *testing.T, clients *test.Clients, hr *gatewayv1alpha1.HTTPRouteSpec, name string) (*gatewayv1alpha1.Gateway, context.CancelFunc) {
+	t.Helper()
+
+	listeners := make([]gatewayv1alpha1.Listener, 0, len(hr.Hostnames))
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{
+		httpRouteNameLabelKey: name,
+	}}
+
+	for _, host := range hr.Hostnames {
+		host := host
+		route := gatewayv1alpha1.RouteBindingSelector{
+			Namespaces: &gatewayv1alpha1.RouteNamespaces{
+				From: routeSelectTypePtr(gatewayv1alpha1.RouteSelectAll),
+			},
+			Selector: &labelSelector,
+			Kind:     "HTTPRoute",
+		}
+
+		listeners = append(listeners, gatewayv1alpha1.Listener{
+			Hostname: &host,
+			Port:     gatewayv1alpha1.PortNumber(80),
+			Protocol: gatewayv1alpha1.HTTPProtocolType,
+			Routes:   route})
+	}
+
+	var address []gatewayv1alpha1.GatewayAddress
+	var gatewayClass string
+	if isClusterLocal(hr.Hostnames) {
+		address = []gatewayv1alpha1.GatewayAddress{{Type: gatewayAddressTypePtr(gatewayv1alpha1.NamedAddressType), Value: test.LocalGatewayAddress()}}
+		gatewayClass = test.LocalGatewayClass()
+	} else {
+		address = []gatewayv1alpha1.GatewayAddress{{Type: gatewayAddressTypePtr(gatewayv1alpha1.NamedAddressType), Value: test.GatewayAddress()}}
+		gatewayClass = test.GatewayClass()
+	}
+
+	gw := &gatewayv1alpha1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: gatewayv1alpha1.GatewaySpec{
+			GatewayClassName: gatewayClass,
+			Listeners:        listeners,
+			Addresses:        address,
+		},
+	}
+
+	t.Cleanup(func() { clients.GatewayAPIClient.Gateways.Delete(ctx, gw.Name, metav1.DeleteOptions{}) })
+	if err := reconciler.RetryTestErrors(func(attempts int) (err error) {
+		if attempts > 0 {
+			t.Logf("Attempt %d creating gateway %s", attempts, gw.Name)
+		}
+		gw, err = clients.GatewayAPIClient.Gateways.Create(ctx, gw, metav1.CreateOptions{})
+		if err != nil {
+			t.Logf("Attempt %d creating gateway failed with: %v", attempts, err)
+		}
+		return err
+	}); err != nil {
+		t.Fatalf("Error creating Gateway %q: %v", gw.Name, err)
+	}
+
+	return gw, func() {
+		err := clients.GatewayAPIClient.Gateways.Delete(ctx, gw.Name, metav1.DeleteOptions{})
+		if err != nil {
+			t.Errorf("Error cleaning up Gateway %q: %v", gw.Name, err)
+		}
+	}
+}
+
+// CreateHTTPRoute creates a HTTPRoute resource with a corresponding Gateway.
 func CreateHTTPRoute(ctx context.Context, t *testing.T, clients *test.Clients, spec gatewayv1alpha1.HTTPRouteSpec, io ...Option) (*gatewayv1alpha1.HTTPRoute, context.CancelFunc) {
 	t.Helper()
 
 	name := test.ObjectNameForTest(t)
 
-	// Create a simple HTTPRoute over the Service.
+	gateway, cancel := CreateGateway(ctx, t, clients, &spec, name)
```

Comment:
I do not think its a good practice to have 1 Gateway per route. In some implementations that will actually provision a real load balancer (with real cost, slowness, etc)

Reply:
Thank you. Hmm... Gateway API raises the 1:1 example. But some implementation does not want us to use it (?).

https://gateway-api.sigs.k8s.io/concepts/api-overview/#route-binding
> One-to-one - A Gateway and Route may be deployed and used by a single owner and have a one-to-one relationship. Team C is an example of this.

Reply:
I think in that example they are deploying one physical load balancer (ie envoy in many cases) for each gateway though. So you CAN have 1:1, but then you are going to end up with 1000s of envoys.

My understanding was Contour deploys an Envoy per? I may be wrong though. At least for Istio, we would still require all the gateways to be in the same namespace

Reply:
BTW, since https://github.com/kubernetes-sigs/gateway-api/pull/725/files has landed, the selection of Route<->Gateway has completely changed. Its now ~identical to how Istio GW/VS works

Reply:
Okay, I will try to ask about it in Gateway API.
As you said, if some implementations will provision a real load balancer, it definitely needs a real cost a lot.

Reply:
Okay, I will try to ask about it in Gateway API.
As you said, if some implementations will provision a real load balancer, it definitely needs a real cost a lot.